src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs

```diff
@@ -0,0 +1,306 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Drawing.Drawing2D;
+using System.IO;
+using System.IO.Pipes;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.RemoteExecutor;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Sockets.Tests
+{
+    // Test cases for DuplicateAndClose, Socket(socketInformation), Socket.UseOnlyOverlappedIO,
+    // and asynchronous IO behavior for duplicate sockets.
+    public class SocketDuplicationTests
+    {
+        private readonly ArraySegment<byte> _receiveBuffer = new ArraySegment<byte>(new byte[32]);
+        const string TestMessage = "test123!";
+        private static ArraySegment<byte> TestBytes => Encoding.ASCII.GetBytes(TestMessage);
+        private static string GetMessageString(ArraySegment<byte> data, int count) =>
+            Encoding.ASCII.GetString(data.AsSpan().Slice(0, count));
+        private static int CurrentProcessId => Process.GetCurrentProcess().Id;
+
+
+        [Fact]
+        public void UseOnlyOverlappedIO_AlwaysFalse()
+        {
+            using Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+
+            Assert.False(s.UseOnlyOverlappedIO);
+            s.UseOnlyOverlappedIO = true;
+            Assert.False(s.UseOnlyOverlappedIO);
+        }
+
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [Fact]
+        public void DuplicateAndClose_TargetProcessDoesNotExist_Throws_SocketException()
+        {
+            using Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+
+            Assert.Throws<SocketException>(() => socket.DuplicateAndClose(-1));
+        }
+
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [Fact]
+        public void DuplicateAndClose_WhenDisposed_Throws()
+        {
+            Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            socket.Dispose();
+
+            Assert.Throws<ObjectDisposedException>(() => socket.DuplicateAndClose(CurrentProcessId));
+        }
+
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void BlockingState_IsTransferred(bool blocking)
+        {
+            using Socket original = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp)
+            {
+                Blocking = blocking
+            };
+            Assert.Equal(blocking, original.Blocking);
+
+            SocketInformation info = original.DuplicateAndClose(CurrentProcessId);
+
+            using Socket clone = new Socket(info);
+            Assert.Equal(blocking, clone.Blocking);
+        }
+
+        public class NotSupportedOnUnix
```

Comment:
It's better not to nest one test class into another. It impairs readability.

Reply:
A [hierarchical pattern](https://haacked.com/archive/2012/01/02/structuring-unit-tests.aspx/) seems actually more readable for me in many cases: helps avoiding long test names, organizing stuff into groups instead. However in this particular case it would be definitely better to ungroup them.

I wonder if we have a written guideline for test code?

Reply:
I don't think I've seen this pattern elsewhere in our tests.

Our unit testing is modeled after a book... I believe The Art of Unit Testing. @davidsh can confirm.